File: 프로그래머스/2/172927. 광물 캐기/광물 캐기.cpp
#include <bits/stdc++.h>
using namespace std;

int calc(int pick, string mineral){
    if(pick == 0) return 1;
    else if(pick == 1){
        if(mineral == "diamond") return 5;
        else return 1;
    }else{
        if(mineral == "diamond") return 25;
        else if( mineral == "iron") return 5;
        else return 1;
    }
}

void DFS(vector<int>& picks, vector<string>& minerals, int& ans, int sum, int pos){
    if((picks[0] == 0 && picks[1] == 0 && picks[2] == 0) || pos == minerals.size()){
        ans = min(ans, sum);
        return;
    }

    for(int i = 0; i<3; i++){
        if(picks[i] != 0){
            picks[i]--;
            int tmpSum = sum;
            int tmpPos = pos;

            for(;tmpPos < pos + 5 && tmpPos < minerals.size(); tmpPos++){
                tmpSum += calc(i, minerals[tmpPos]);
            }
            DFS(picks, minerals, ans, tmpSum, tmpPos);
            picks[i]++;
        }
    }
}

int solution(vector<int> picks, vector<string> minerals) {
    int answer = INT_MAX;

    DFS(picks, minerals, answer, 0,0);
    return answer;
}

File: 프로그래머스/2/172927. 광물 캐기/광물 캐기.js
function solution(picks, minerals) {
  var answer = 0;
  let total = picks[0] + picks[1] + picks[2];

  let simulated = [];
  for(let i = 0; i<total*5; i+=5){
    let tmp = [0,0,0];

    for(let j = i; j<Math.min(i+5, minerals.length); j++){
      if(minerals[j] === "diamond"){
        tmp[0]++;
        tmp[1]+=5;
        tmp[2]+=25;
      }else if(minerals[j] === "iron"){
        tmp[0]++;
        tmp[1]++;
        tmp[2]+=5;
      }else{
        tmp[0]++;
        tmp[1]++;
        tmp[2]++;
      }
    }
    simulated.push([...tmp]);
  }

  simulated.sort((a,b) => b[2] - a[2]);

  for(let i = 0; i<simulated.length; i++){
    if(picks[0]){
      picks[0]--;
      answer += simulated[i][0];
    }else if(picks[1]){
      picks[1]--;
      answer += simulated[i][1];
    }else if(picks[2]){
      picks[2]--;
      answer += simulated[i][2];
    }else break;

    if((i + 1) * 5 >= minerals.length) break;
  }



  return answer;
}